src/DI/ContainerBuilder.php

```diff
@@ -55,16 +55,17 @@ public function __construct()
 
 	/**
 	 * Adds new service definition.
-	 * @return Definitions\ServiceDefinition
+	 * @return Definition|Definitions\ServiceDefinition
 	 */
 	public function addDefinition(?string $name, Definition $definition = null): Definition
 	{
 		$this->needsResolve = true;
+		$nameType = $name === null ? null : key([$name => 1]); /** @var string|int|null $nameType */
 		if ($name === null) {
 			for ($i = 1; isset($this->definitions['0' . $i]) || isset($this->aliases['0' . $i]); $i++);
 			$name = '0' . $i; // prevents converting to integer in array key
 
-		} elseif (is_int(key([$name => 1])) || !preg_match('#^\w+(\.\w+)*$#D', $name)) {
+		} elseif (is_int($nameType) || !preg_match('#^\w+(\.\w+)*$#D', $name)) {
```

Comment:
@dg This `key()` usage seems like hacky way to check if string is numeric and I think auto-conversion will stop working with some strict directive in future php versions. Maybe `Nette\Utils\Validators::isNumeric|isNumericInt()` would work here.

Reply:
Yes, it is intentional.

Reply:
Maybe it should be fixed in PhpStan, key() can return int.

Reply:
It can, but I think problem here is in `key()` used to convert numeric value to int, it's a side-effect.

Reply:
It's also the surest way to verify that the key will not be number.

Reply:
`key()` is definitely bug in phpstan, caused by combination of https://github.com/phpstan/phpstan/blob/3ef2c05f0fc5dbe413674ef33cbbd60c33bb1ce7/src/Type/Php/ArrayKeyDynamicReturnTypeExtension.php and `ArrayType::getIterableKeyType()` being incorrect because it prefer usefulness over accuracy.

Reply:
I suggest fix it like in `Nette\Utils\Validators::isNumericInt()`, specifically: `($name !== null && preg_match('#^[+-]?\d+\z#', $name))`

Reply:
Ok, I am very sorry. Correct pattern is `preg_match('#^[+-]?([1-9]\d*|0)\z#', $name)`.